### DISCUSSION TITLE
display end experiment message for more than a flash

### DISCUSSION THREAD

**Discussion by Kedersha** [jspsych/jsPsych#2836] [Q&A]
ETA: please disregard, I think I've figured this out; see discussion on the JATOS forum.

(Also posted on JATOS forum [here](https://forum.cogsci.nl/discussion/8216/display-endpage-html-for-more-than-a-flash#latest).)

Hi! I'm writing a study in jsPsych and JATOS, and at several points we want to be able to end the study based on users' responses (failing to consent, sharing information that makes them ineligible to participate). JsPsych and JATOS both theoretically have methods for displaying a page after the study is over, explaining to participants why they've been abruptly booted.

However, the simple html page I wrote (endPage.html, according to what JATOS docs suggested) just flashes on the screen for less than a second before the tab closes; the same happens if I use jsPsych.endExperiment() to show that same message.

Is there a way to keep a message up until the user closes the tab? Have you seen this behaviour before?

Relevant code snippets... JATOS endStudy calls:

```
jatos.onLoad(function(){

var jsPsych = initJsPsych({
      on_close:  function() {
        var resultData = jsPsych.data.get().ignore('internal_node_id');
        jatos.endStudy(resultData.csv());
      },
      on_finish: function() {
        var resultData = jsPsych.data.get().ignore('internal_node_id');
        jatos.submitResultData(resultData.csv(),jatos.endStudy);
      }
    });
```

A sample trial that ends study early:

```
var info = {
    type: jsPsychHtmlButtonResponse,
    stimulus: '<h2>Participant Consent</h2><p>Please download and read <a href="CONSENT.pdf" target="_blank">this consent form</a>.</p><p>By continuing, you are consenting to this study.</p>',
    choices: ['Could not download','Agree to participate'],
    on_finish: function(data) {
      if(data.response == 0){
        jsPsych.endExperiment(endmessage,data);
      }
    }
}
```

**Comment by becky-gilbert** [comment]
Hi @Kedersha!
I'm not sure about the `jsPsych.endExperiment` solution, but I had the same problem when I used `jatos.endStudy`. The problem is that if you tell JATOS to end the study with this function, by default it will immediately redirect to a different page. 
My solution was to use the [`jatos.endStudyAjax` function](https://www.jatos.org/jatos.js-Reference.html#jatosendstudyajax), which marks the study as ended in JATOS but without any redirection. (BTW It looks like the JATOS developers have updated the [`jatos.endStudy` function](https://www.jatos.org/jatos.js-Reference.html#jatosendstudy) so that now it has a `showEndPage` parameter that you can set to `false`, so that should do the same thing).
I hope that helps! Let us know 😄

**Reply by becky-gilbert** [reply]
Sorry, I just realized my answer wasn't very clear. What I would do is leave everything in `initJsPsych` the way it is, but add a [conditional node](https://www.jspsych.org/7.3/overview/timeline/#conditional-timelines) after your `info` trial that checks the response from the last trial, and if the response = 0, then run a timeline that just contains the end message. And this is the point where you'd also want to send the data back to JATOS and use `jatos.endStudyAjax` (or `jatos.endStudy` with `showEndPage=false`). So it might look something like this:
```js
var info = {
    type: jsPsychHtmlButtonResponse,
    stimulus: '<h2>Participant Consent</h2><p>Please download and read <a href="CONSENT.pdf" target="_blank">this consent form</a>.</p><p>By continuing, you are consenting to this study.</p>',
    choices: ['Could not download','Agree to participate']
};
experiment_timeline.push(info);

var end_message = {
  // create your end message trial here,
  // e.g. using html-keyboard-response and no key choices
}; 

var end_conditional = {
  timeline: [end_message],
  conditional_function: function() {
    var data = jsPsych.data.get().last(1).values()[0];
    if (data.response == 0) {
      // send the data back and end the study (without redirecting)
      var resultData = jsPsych.data.get().ignore('internal_node_id').csv();
      jatos.submitResultData(resultData, jatos.endStudyAjax);
      // show the end_message trial
      return true;  
    } else {
      // skip the end_message trial
      return false ; 
    }
  }
};
experiment_timeline.push(end_conditional);
```
